### PR TITLE
Uda udf test

### PR DESCRIPTION
**Pull request by KnifeyMoloko**
This commit introduces a feature test in SCT for User Defined
Functions' and User Defined Aggregates' basic functionality,
i.e. creating UDFs, incorporating them into UDAs and using
both in CQL queries. The functions used are implemented in
both Lua and WASM (by compiling C programs into WASM modules
and then WAT human-readable scripts). All the functions used
are very basic (sum, division, length-of-text etc.)
The test itself pre-creates the UDFs and UDAs before using
cassandra-stress with a custom user profile to run queries
using the pre-created UDFs and UDAs alongside a write workload.

Jenkins job for testing the changes: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/uda-udf/

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: uda_udf_test.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            c3_value = row_result.c3
+            c7_value = row_result.c7
+            assert row_query.verifier_func(c2_value, c3_value, c7_value), \
+                "Expected row values to not be None, at least one them was. " \
+                "c2_value: %s, c3_value: %s, c7 value: %s" % (c2_value, c3_value, c7_value)
+
+            for verification in verifications:
+                self.log.info("Running UDF verification: %s; query: %s", verification.name, verification.query)
+                query_result = session.execute(verification.query).one()
+                assert verification.verifier_func(c2_value, c3_value, c7_value, query_result)
```
Comment (soyacz) [inline uda_udf_test.py:105]: how can you be sure that each verification query returns the same row as in `row_result` when there's no `where` clause in queries?

File: uda_udf_test.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        for udf in UDFS.values():
+            self.log.info("Creating the following UDF: %s", udf.name)
+            cmd = udf.get_create_query(ks="ks")
+            with self.db_cluster.cql_connection_patient(node=node) as session:
+                session.execute(cmd)
+            self.log.info("UDF %s created", udf.name)
+
+        self._verify_udf_functions()
+
+        for uda in UDAS.values():
```
Comment (soyacz) [inline uda_udf_test.py:38]: please add log.info that UDA is created similary to UDF for better understanding.
Also, why there's no UDA verification like in UDF?

File: uda_udf_test.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                                  f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                            verifier_func=lambda c2, c3, c7, query_response: c2 + c3 == query_response.result),
+            UDFVerification(name="xwasm_div",
+                            query=f"SELECT {self.KEYSPACE_NAME}.xwasm_div(c2, c3) AS result "
+                                  f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                            verifier_func=lambda c2, c3, c7, query_response: c2 // c3 == query_response.result),
+            # UDFVerification(name="xwasm_fib",
+            #                 query=f"SELECT {self.KEYSPACE_NAME}.xwasm_fib(c2) AS result "
+            #                       f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+            #                 verifier_func=lambda c2, c3, c7, query_response: self._fib(c2) == query_response.result)
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply uda_udf_test.py]: I decided to leave the fibonacci function out of the test, since it can be pretty resource consuming and I'm not using it in the aggregate either way.

File: sdcm/utils/udf_scripts/xwasm_fib.yaml [hunk 4]
```diff
@@ -1,162 +1,48 @@
 # UDF script: returns the string length of the provided text var
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply sdcm/utils/udf_scripts/xwasm_fib.yaml]: Leftover from the text length counter function - fixed (I removed fib altogether)

File: uda_udf_test.py (hunk 1 above)
Comment (KnifeyMoloko) [reply uda_udf_test.py:100]: Why would it not? The data is pre-written and doesn't change during the test - I'd expect the results to be consistent in such a situation. (Plus, the verifications are passing in my staging tests.)

File: data_dir/c-s_uda_udf.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  CREATE TABLE uda_udf (
+        pk timeuuid,
+        ck int,
+        c1 timeuuid,
+        c2 int,
+        c3 int,
+        c4 bigint,
+        c5 double,
+        c6 float,
+        c7 text,
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply data_dir/c-s_uda_udf.yaml]: Good idea - I added the type subscripts.

File: uda_udf_test.py (hunk 2 above)
Comment (KnifeyMoloko) [reply uda_udf_test.py:38]: Added. And now there is a UDA verification, that's a good idea, I originally missed.

File: uda_udf_test.py (hunk 1 above)
Comment (soyacz) [reply uda_udf_test.py:101]: I meant it is not sorted with `WHERE` or `ORDER BY` - so there's no certainty it will result in the same row. I wonder how it works.

Comment (KnifeyMoloko) [reply uda_udf_test.py:107]: The data is already sorted by partition (pk) and cluster keys (ck_int). From our docs[1] on the `CLUSTERING ORDER` param for table creation: `it also changes the default order of results when queried (if no ORDER BY is provided). Results are always returned in clustering order (within a partition).` In this case we stick to defaults, no CLUSTERING ORDER is defined, so the data will always be retrieved in the same order while there is nothing affecting the partitioning and clustering. 
Either way, I also double checked on a small toy example with creating different client session and redoing a `SELECT * LIMIT 1;` query and the results are stable.

[1] https://docs.scylladb.com/stable/cql/ddl.html#create-table-statement

Comment (soyacz) [reply uda_udf_test.py:113]: Thanks for the explanation @KnifeyMoloko

**Review by soyacz — DISMISSED** [review]

**Comment by roydahan** [comment]
@eliransin please review or assign someone to review.

**Comment by KnifeyMoloko** [comment]
Adding @wmitros as he know the feature best. @wmitros could you take a look at this PR? It's for the UDF and UDA features - creating, combining into a UDA and running as part of a cassandra-stress workload.

File: sdcm/utils/udf.py [hunk 6]
```diff
@@ -49,7 +49,7 @@ class UDF(BaseModel):
     script: str
 
     def get_create_query(self, ks: str, create_or_replace: bool = True) -> str:
-        create_part = "CREATE OR REPLACE" if create_or_replace else "CREATE"
+        create_part = "CREATE OR REPLACE FUNCTION" if create_or_replace else "CREATE"
```
Comment (fgelcer) [inline sdcm/utils/udf.py:52]: shouldn't it be?
```
"CREATE OR REPLACE FUNCTION" if create_or_replace else "CREATE FUNCTION"
```

File: uda_udf_test.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                                                     logdir=self.logdir)
+
+    def test_uda_and_udf(self) -> None:
+        self.log.info("Starting UDA/UDF test...")
+        self.prewrite_db_with_data()
+        node: BaseNode = self.db_cluster.get_node()
+
+        for udf in UDFS.values():
+            self.log.info("Creating the following UDF: %s", udf.name)
+            cmd = udf.get_create_query(ks="ks")
```
Comment (fgelcer) [inline uda_udf_test.py:37]:
```suggestion
            cmd = udf.get_create_query(ks=self.KEYSPACE_NAME)
```

File: uda_udf_test.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            with self.db_cluster.cql_connection_patient(node=node) as session:
+                session.execute(cmd)
+            self.log.info("UDF %s created.", udf.name)
+
+        self._verify_udf_functions()
+
+        for uda in UDAS.values():
+            cmd = uda.get_create_query_string(ks="ks")
```
Comment (fgelcer) [inline uda_udf_test.py:45]:
```suggestion
            cmd = uda.get_create_query_string(ks=self.KEYSPACE_NAME)
```

**Review by fgelcer — COMMENTED** [review]
LGTM, although have few small comments

File: test-cases/features/uda_udf.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+test_duration: 90
+
+prepare_write_cmd: "cassandra-stress user profile=/tmp/c-s_uda_udf.yaml ops'(insert=1)' cl=QUORUM duration=2m -mode native cql3 -rate threads=10"
+stress_cmd: ["cassandra-stress user profile=/tmp/c-s_uda_udf.yaml ops'(my_avg=1, lua_var_length_counter=33, xwasm_plus=33, xwasm_div=32)' cl=QUORUM duration=30m -mode native cql3 -rate threads=20"]
```
Comment (roydahan) [inline test-cases/features/uda_udf.yaml:4]: It should be a longevity test, much longer with much more data in order to exercise the queries on larger partitions, and much more rows.

File: test-cases/features/uda_udf.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+n_db_nodes: 3
+n_loaders: 1
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
```
Comment (roydahan) [inline test-cases/features/uda_udf.yaml:10]: Please change according to the dataset and use i3en or i4i.

File: test-cases/features/uda_udf.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+append_scylla_yaml: |
+  experimental_features:
+     - udf
+  enable_user_defined_functions: true
+  user_defined_function_time_limit_ms: 10000
+  user_defined_function_allocation_limit_bytes: 1048576
```
Comment (roydahan) [inline test-cases/features/uda_udf.yaml:16]: Do we exercise or test these parameters somewhere?

**Review by roydahan — CHANGES_REQUESTED** [review]
in general looks good.
Just need to change the test to be longer and harder for scylla.

File: data_dir/c-s_uda_udf.yaml [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        ck_int int,
+        c1_timeuuid timeuuid,
+        c2_int int,
+        c3_int int,
+        c4_bigint bigint,
+        c5_double double,
+        c6_float float,
+        c7_text text,
+        PRIMARY KEY(pk, ck_int)
+  ) WITH compaction = { 'class':'SizeTieredCompactionStrategy' }
```
Comment (yarongilor) [inline data_dir/c-s_uda_udf.yaml:24]: why specifying 'SizeTieredCompactionStrategy' and not leaving the default, so for enterprise it'll use ICS?

File: uda_udf_test.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+        self._verify_uda_aggregates()
+
+        uda_udf_thread = self.run_uda_udf_thread()
+
+        # wait for stress to complete
+        self.verify_stress_thread(cs_thread_pool=uda_udf_thread)
+        self.log.info("Test completed")
+
+    def run_uda_udf_thread(self) -> Tuple[CassandraStressThread, CassandraStressThread]:
```
Comment (yarongilor) [inline uda_udf_test.py:57]: Not sure why returned type annotation has 2 threads while returning a single thread

File: test-cases/features/uda_udf.yaml (hunk 9 above)
Comment (KnifeyMoloko) [reply test-cases/features/uda_udf.yaml:4]: I'm currently testing this approach by running the uda/udf after a much longer prepare_write.

File: unit_tests/test_uda.py [hunk 14]
```diff
@@ -85,10 +85,10 @@ def test_get_create_query_string(self):
...
     def test_load_uda_from_yaml(self):
         expected_create_string = "CREATE AGGREGATE testing.my_uda(int) " \
-                                 "SFUNC plus " \
-                                 "STYPE tinyint " \
-                                 "REDUCEFUNC plus " \
-                                 "FINALFUNC fib " \
+                                 "SFUNC xwasm_plus " \
+                                 "STYPE int " \
+                                 "REDUCEFUNC xwasm_plus " \
+                                 "FINALFUNC xwasm_div_fin " \
```
Comment (wmitros) [inline unit_tests/test_uda.py:91]: I see that we're not actually executing this query, but this aggregate would not work because the types of `SFUNC`, `REDUCEFUNC` and `FINALFUNC` don't match.
If we wanted this UDA to be a sum, we would need the `xwasm_div_fin` to be an identity function instead, and if we wanted it to be an average, we would need to change the `SFUNC` to `xwasm_plus_acc` and the REDUCEFUNC to an UDF that sums two accumulator tuples.

**Review by wmitros — COMMENTED** [review]
Looks good, but testing WASM UDFs for performance will require more work.

The issue lies in the way we're passing arguments and return values between scylla and the wasm runtime. When all of them are WASM native types (basically ints and floats) there is no problem - we can test their performance already (for example, a SUM aggregate)

The problem arises when we're using any more complicated types (like a tuple as the accumulator). In these cases, the value needs to be serialized and stored in the WASM memory. This is problematic, because is no proper memory allocator in available WASM by default. It can be supplied by the user, but by default in scylla, we aren't doing that, and we allocate the memory for arguments super inefficiently.

To actually be efficient, we need to use [`_scylla_abi`](https://github.com/scylladb/scylladb/blob/master/docs/dev/wasm.md#memory-management) of `2` instead of `1`, and provide an allocator in the WASM bytecode ourselves. The easiest way to achieve this for me was by compiling to the `wasm32-wasi` target, but it is also possible to achieve simply #including some `malloc` implementation.

Another way of achieving better performance is to use the Rust helper [`library`](https://github.com/wmitros/scylla-rust-udf/tree/initials) that we're developing. The library is not yet ready though, and we would need someone knowing some Rust for that. 

Considering this, I suggest we postpone performance testing WASM UDFs with complex types until the library is ready. They should be easy enough to add with the context that this patch introduces. 

It's also worth noting that because https://github.com/scylladb/scylladb/pull/11351 was recently merged, there are now more configuration options (specifically https://github.com/scylladb/scylladb/pull/11351/commits/b8d28a95bf03945a7e6e7ce5108d1a21e56b6547) . I'm not saying they should be immediately used in this patch, but in particular `wasm_udf_yield_fuel` and `wasm_udf_memory_limit` correlate (though not exactly one-to-one) to the `user_defined_function_time_limit_ms` and `user_defined_function_allocation_limit_bytes` that I see we're using.

By the way, thanks for working on this, this feature really needed some (performance) testing. Also, sorry for reviewing this so late

File: unit_tests/test_uda.py (hunk 14 above)
Comment (KnifeyMoloko) [reply unit_tests/test_uda.py]: Good catch. Even if this is not executed, better for it to not be confusing. I changed it to include a `xwasm_simple_return_int` function which is just a `int` identity function.

File: sdcm/utils/udf_scripts/xwasm_plus_acc.yaml [hunk 15]
```diff
@@ -0,0 +1,151 @@
+# UDF script: returns a tuple with the accumulator incremented by arg and counter incremented by 1
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply sdcm/utils/udf_scripts/xwasm_plus_acc.yaml:1]: Agreed. I'm considering removing this function (and the final div function in xwasm as well), since they don't perform tuple deserialization like they should, so no changes to this one atm, but I'll be adding the source code and compile commands to other added scripts.

File: sdcm/utils/udf.py (hunk 6 above)
Comment (KnifeyMoloko) [reply sdcm/utils/udf.py:52]: Fixed - corrected

File: test-cases/features/uda_udf.yaml (hunk 10 above)
Comment (KnifeyMoloko) [reply test-cases/features/uda_udf.yaml]: Fixed

File: test-cases/features/uda_udf.yaml (hunk 11 above)
Comment (KnifeyMoloko) [reply test-cases/features/uda_udf.yaml:16]: We do, in the sense that setting these values low will end with errors being thrown. IIRC unit tests check for boundary values, here I just wanted to make sure that if the UDF failed, it's not because of a user config error.

File: data_dir/c-s_uda_udf.yaml (hunk 12 above)
Comment (KnifeyMoloko) [reply data_dir/c-s_uda_udf.yaml]: I'm actually not sure why it was set that way. Either way, I got rid of it -we can go with defaults here.

File: uda_udf_test.py (hunk 13 above)
Comment (KnifeyMoloko) [reply uda_udf_test.py]: Good catch - fixed

File: uda_udf_test.py (hunk 8 above)
Comment (KnifeyMoloko) [reply uda_udf_test.py]: Fixed

File: uda_udf_test.py (hunk 7 above)
Comment (KnifeyMoloko) [reply uda_udf_test.py]: Fixed

File: sdcm/utils/udf_scripts/xwasm_plus_acc.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (roydahan) [reply sdcm/utils/udf_scripts/xwasm_plus_acc.yaml:1]: @KnifeyMoloko maybe what we should do in SCT is to have the C++ code and the SCT will be responsible of the compilation + translation to wat format.
So, we can change the human readable code only.

**Review by roydahan — APPROVED** [review]

**Review by soyacz — APPROVED** [review]
LGTM.
@KnifeyMoloko any more fixes to this needed? Merge?

**Comment by fgelcer** [comment]
> LGTM. @KnifeyMoloko any more fixes to this needed? Merge?

@soyacz AFAICT there will be more work afterwards to enlarge the testing coverage (like adding in some longevity+ one or more of these UDA/UDF user profiles), but unrelated to this current PR, and IMHO it can be merged.

**Comment by soyacz** [comment]
@KnifeyMoloko please set backport labels

**Comment by fgelcer** [comment]
@roydahan , do we need to backport this one?

**Comment by roydahan** [comment]
No, but we will Fast-Forward 5.2 so we will get it anyway there.
